### PR TITLE
M3-609 View and update account contact information.

### PR DESCRIPTION
**Pull request by rjstires**
## Purpose
Allow users to update their billing contact details.

## Known Issues
- State / Province is not a select.
Country select is mildly slow, but not game changing. We can convert after the virtualization PR gets - merged.

### CODE REVIEW

File: src/features/Account/AccountPanels/UpdateContactInformationPanel/UpdateContactInformationPanel.tsx [hunk 1]
```diff
@@ -1,29 +1,635 @@
...
+      loading, errors, data, lastUpdated,
+    } = this.props;
+    return (
+      <ExpansionPanel
+        heading="Update Contact Information"
+        actions={this.renderFormActions}
+      >
+        {
+          loading && lastUpdated === 0
+            ? this.renderLoading()
```
Comment (martinmckenna) [inline src/features/Account/AccountPanels/UpdateContactInformationPanel/UpdateContactInformationPanel.tsx]: Not a fan of this triple nested ternary. Can we abstract this elsewhere?

File: src/features/Account/AccountPanels/UpdateContactInformationPanel/UpdateContactInformationPanel.tsx [hunk 2]
```diff
@@ -1,29 +1,635 @@
...
+
+    const generalError = hasErrorFor('none');
+
+    return (
+      <Grid container>
+        {generalError && <Grid item xs={12}><Notice error text={generalError} /></Grid>}
+        {success && <Grid item xs={12}><Notice success text={success} /></Grid>}
+
+        <Grid item xs={12}>
+          <TextField
```
Comment (martinmckenna) [inline src/features/Account/AccountPanels/UpdateContactInformationPanel/UpdateContactInformationPanel.tsx]: Every one of these textfields re-renders whenever you change one. I think at the very least, we can wrap them in a renderGuard. I'm getting really bad performance issues when selecting a country.

File: src/features/Account/AccountPanels/UpdateContactInformationPanel/UpdateContactInformationPanel.tsx [hunk 3]
```diff
@@ -1,29 +1,635 @@
...
+  };
+
+  updateCompany = (e: React.ChangeEvent<HTMLInputElement>) => {
+    this.composeState([
+      set(L.fields.company, e.target.value),
+    ])
+  };
+
+  updateCountry = (e: React.ChangeEvent<HTMLSelectElement>) => {
+    console.log('updateCountry');
```
Comment (martinmckenna) [inline src/features/Account/AccountPanels/UpdateContactInformationPanel/UpdateContactInformationPanel.tsx]: ahhhhh

File: src/features/Account/AccountPanels/UpdateContactInformationPanel/UpdateContactInformationPanel.tsx [hunk 4]
```diff
@@ -1,29 +1,635 @@
...
+    this.composeState([
+      set(L.fields.country, e.target.value),
+    ], () => console.log(this.state.fields.country))
```
Comment (martinmckenna) [inline src/features/Account/AccountPanels/UpdateContactInformationPanel/UpdateContactInformationPanel.tsx]: console log

File: src/features/Account/AccountPanels/UpdateContactInformationPanel/UpdateContactInformationPanel.tsx [hunk 5]
```diff
@@ -1,29 +1,635 @@
...
+
+  renderFormActions = () => {
+    const { loading, lastUpdated, errors } = this.props;
+
+    if ((loading && lastUpdated === 0) || errors) { return null; }
+
+    return (
+      <ActionsPanel>
+        <Button type="primary" onClick={this.submitForm} loading={this.state.submitting}>Save</Button>
+        <Button type="secondary" onClick={this.resetForm}>Reset</Button>
```
Comment (martinmckenna) [inline src/features/Account/AccountPanels/UpdateContactInformationPanel/UpdateContactInformationPanel.tsx]: We should decide what we want to name these buttons. I like "Reset", but we've historically been using "Cancel" as the reset button text

File: src/features/Account/AccountPanels/UpdateContactInformationPanel/UpdateContactInformationPanel.tsx [hunk 6]
```diff
@@ -1,29 +1,635 @@
...
+            label="Company"
+            value={defaultTo(account.company, fields.company)}
+            errorText={hasErrorFor('company')}
+            onChange={this.updateCompany}
+          />
+        </Grid>
+
+        <Grid item xs={6}>
+          <TextField
+            label="Email"
```
Comment (martinmckenna) [inline src/features/Account/AccountPanels/UpdateContactInformationPanel/UpdateContactInformationPanel.tsx]: `type` prop of `email` for mobile

File: src/features/Account/AccountPanels/UpdateContactInformationPanel/UpdateContactInformationPanel.tsx [hunk 7]
```diff
@@ -1,29 +1,635 @@
...
+            value={defaultTo(account.email, fields.email)}
+            errorText={hasErrorFor('email')}
+            onChange={this.updateEmail}
+          />
+        </Grid>
+
+        <Grid item xs={6}>
+          <TextField
+            label="Phone Number"
```
Comment (martinmckenna) [inline src/features/Account/AccountPanels/UpdateContactInformationPanel/UpdateContactInformationPanel.tsx]: `type` prop of either `tel` or `number`. I think `number` is more likely to show a number pad on mobile

File: src/features/Account/AccountPanels/UpdateContactInformationPanel/UpdateContactInformationPanel.tsx [hunk 8]
```diff
@@ -1,29 +1,635 @@
...
+            onChange={this.updateCountry}
+            select
+          >
+            {UpdateContactInformationPanel.countryItems()}
+          </TextField>
+        </Grid>
+
+        <Grid item xs={6}>
+          <TextField
+            label="Zip / Postal Code"
```
Comment (martinmckenna) [inline src/features/Account/AccountPanels/UpdateContactInformationPanel/UpdateContactInformationPanel.tsx]: again, should probably have a `number` type

File: src/features/Account/AccountPanels/UpdateContactInformationPanel/UpdateContactInformationPanel.tsx [hunk 9]
```diff
@@ -1,29 +1,635 @@
...
+};
 
 class UpdateContactInformationPanel extends React.Component<CombinedProps, State> {
-  state: State = {};
+  state: State = {
+    submitting: false,
+    fields: {},
+  };
+
+  composeState = (
```
Comment (asauber) [inline src/features/Account/AccountPanels/UpdateContactInformationPanel/UpdateContactInformationPanel.tsx]: This is the third time that this function's been defined. Seems like it's time to factor it out.

File: src/features/Account/AccountPanels/UpdateContactInformationPanel/UpdateContactInformationPanel.tsx [hunk 10]
```diff
@@ -1,29 +1,635 @@
...
+import scrollErrorIntoView from 'src/utilities/scrollErrorIntoView';
 
 type ClassNames = 'root';
 
 const styles: StyleRulesCallback<ClassNames> = (theme: Theme) => ({
   root: {},
 });
 
-interface Props {}
+type Map<T> = (t: T) => T;
```
Comment (asauber) [inline src/features/Account/AccountPanels/UpdateContactInformationPanel/UpdateContactInformationPanel.tsx]: This name shadows the type for `Map` https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map

Maybe `MapFn`?

File: src/features/Account/AccountPanels/UpdateContactInformationPanel/UpdateContactInformationPanel.tsx (hunk 2 above)
Comment (rjstires) [reply src/features/Account/AccountPanels/UpdateContactInformationPanel/UpdateContactInformationPanel.tsx]: The selection of the country is related to MUI select performance (as discussed in the virtualization).

File: src/features/Account/AccountPanels/UpdateContactInformationPanel/UpdateContactInformationPanel.tsx (hunk 3 above)
Comment (rjstires) [reply src/features/Account/AccountPanels/UpdateContactInformationPanel/UpdateContactInformationPanel.tsx]: danke

File: src/features/Account/AccountPanels/UpdateContactInformationPanel/UpdateContactInformationPanel.tsx (hunk 4 above)
Comment (rjstires) [reply src/features/Account/AccountPanels/UpdateContactInformationPanel/UpdateContactInformationPanel.tsx]: Danke

File: src/features/Account/AccountPanels/UpdateContactInformationPanel/UpdateContactInformationPanel.tsx (hunk 8 above)
Comment (rjstires) [reply src/features/Account/AccountPanels/UpdateContactInformationPanel/UpdateContactInformationPanel.tsx]: You would anger a lot of Canadians. https://en.wikipedia.org/wiki/Postal_codes_in_Canada

Comment (martinmckenna) [reply src/features/Account/AccountPanels/UpdateContactInformationPanel/UpdateContactInformationPanel.tsx]: oooh good call

**Comment by alioso** [comment]
styled and ready to go on my end

**Pull request merged**